api/src/test/java/org/pmiops/workbench/api/ConceptSetsControllerTest.java

```diff
@@ -0,0 +1,482 @@
+    cdrVersion.setName("1");
+    //set the db name to be empty since test cases currently
+    //run in the workbench schema only.
+    cdrVersion.setCdrDbName("");
+    cdrVersion = cdrVersionDao.save(cdrVersion);
+
+    Workspace workspace = new Workspace();
+    workspace.setName(WORKSPACE_NAME);
+    workspace.setNamespace(WORKSPACE_NAMESPACE);
+    workspace.setDataAccessLevel(DataAccessLevel.PROTECTED);
+    workspace.setResearchPurpose(new ResearchPurpose());
+    workspace.setCdrVersionId(String.valueOf(cdrVersion.getCdrVersionId()));
+
+    Workspace workspace2 = new Workspace();
+    workspace2.setName(WORKSPACE_NAME_2);
+    workspace2.setNamespace(WORKSPACE_NAMESPACE);
+    workspace2.setDataAccessLevel(DataAccessLevel.PROTECTED);
+    workspace2.setResearchPurpose(new ResearchPurpose());
+    workspace2.setCdrVersionId(String.valueOf(cdrVersion.getCdrVersionId()));
+
+    WorkspacesController workspacesController = new WorkspacesController(workspaceService,
+        cdrVersionDao, cohortDao, userDao, userProvider, fireCloudService, cloudStorageService, CLOCK,
+        userService);
+    stubGetWorkspace(WORKSPACE_NAMESPACE, WORKSPACE_NAME, USER_EMAIL, WorkspaceAccessLevel.OWNER);
+    stubGetWorkspace(WORKSPACE_NAMESPACE, WORKSPACE_NAME_2, USER_EMAIL, WorkspaceAccessLevel.OWNER);
+    workspacesController.createWorkspace(workspace);
+    workspacesController.createWorkspace(workspace2);
+
+    org.pmiops.workbench.firecloud.model.WorkspaceResponse fcResponse =
+        new org.pmiops.workbench.firecloud.model.WorkspaceResponse();
+    fcResponse.setAccessLevel(WorkspaceAccessLevel.OWNER.name());
+    when(fireCloudService.getWorkspace(WORKSPACE_NAMESPACE, WORKSPACE_NAME))
+        .thenReturn(fcResponse);
+    when(fireCloudService.getWorkspace(WORKSPACE_NAMESPACE, WORKSPACE_NAME_2))
+        .thenReturn(fcResponse);
+
+    conceptSetsController = new ConceptSetsController(workspaceService, conceptSetDao, conceptDao,
+        userProvider, CLOCK);
+  }
+
+  @Test
+  public void testGetConceptSetsInWorkspaceEmpty() {
+    assertThat(conceptSetsController.getConceptSetsInWorkspace(WORKSPACE_NAMESPACE, WORKSPACE_NAME)
+        .getBody().getItems()).isEmpty();
+  }
+
+  @Test(expected = NotFoundException.class)
+  public void testGetConceptSetNotExists() {
+    conceptSetsController.getConceptSet(WORKSPACE_NAMESPACE, WORKSPACE_NAME, 1L);
+  }
+
+  @Test(expected = NotFoundException.class)
+  public void testUpdateConceptSetNotExists() {
+    ConceptSet conceptSet = new ConceptSet();
+    conceptSet.setDescription("desc 1");
+    conceptSet.setName("concept set 1");
+    conceptSet.setDomain(Domain.CONDITION);
+    conceptSet.setId(1L);
+    conceptSet.setEtag(Etags.fromVersion(1));
+
+    conceptSetsController.updateConceptSet(WORKSPACE_NAMESPACE, WORKSPACE_NAME, 1L,
+        conceptSet);
+  }
+
+  @Test
+  public void testCreateAndGetConceptSet() {
+    ConceptSet conceptSet = makeConceptSet1();
```

Comment:
So I could do that, but makeConceptSet1() gets used in a bunch of other places, too... would you copy its contents into this new method?

This didn't seem *that* confusing. :)

Reply:
It's not *that* confusing, it just feels like a poor separation.

I'm suggesting: split this into `testCreateConceptSet()` and `testGetConceptSet()`. Move these assertions to the `create` test, and inline the `makeConceptSet1` method (or some variation) there. Continue using the helper from `getConceptSet()` here and elsewhere (but remove all these now duplicated assertions). Other methods can depend on "create" working properly, since there's now independent coverage of that.

True, there could be a bug in the specific helper function, but that's pretty low value coverage IMO.

Reply:
Done.